pgjdbc/src/main/java/org/postgresql/util/NumberParser.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util;
+
+/**
+ * Optimised byte[] to number parser.
+ */
+public class NumberParser {
+  private static final NumberFormatException FAST_NUMBER_FAILED = new NumberFormatException() {
+    @Override
+    public Throwable fillInStackTrace() {
+      return this;
+    }
+  };
+
+  private static final long MAX_LONG_DIV_TEN = Long.MAX_VALUE / 10;
+
+  /**
+   * Optimised byte[] to number parser. This code does not handle null values, so the caller must do
+   * checkResultSet and handle null values prior to calling this function. Fraction part is discarded.
+   *
+   * @param bytes integer represented as a sequence of ASCII bytes
+   * @return The parsed number.
+   * @throws NumberFormatException If the number is invalid or the out of range for fast parsing.
+   *        The value must then be parsed by another (less optimised) method.
+   */
+  public static long getFastLong(byte[] bytes, long minVal, long maxVal) throws NumberFormatException {
+    int len = bytes.length;
+    if (len == 0 || len > 30) {
+      throw FAST_NUMBER_FAILED;
```

Comment:
why 30 ? The largest long is 19 bytes

Reply:
Really 20 for MIN_VALUE ;-)
`
Long.toString(Long.MIN_VALUE).length()
$1 ==> 20
`

As far as I understand the code, there could be representations with period, like "12235345.23424", so I chose this number to cover most of such cases from the one and do not parse very large bytes array from other side.
But it is a weak argument, so I can change it to 20 if will be better.

Reply:
Just scan the string for '.' if it doesn't exist use 20 for negative and 19 for positive

Reply:
Maybe it better to just remove this check?
We already have check `val <= MAX_LONG_DIV_TEN` in the loop and it will work faster(in average) than additional scan.

Reply:
The check is pretty ambiguous so, ya, let's just get rid of it. Optimistically this should not fail like this very often